src/utils/net/download.rs

```diff
@@ -1,202 +1,91 @@
 // Copyright 2019-2023 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use std::{
-    io::SeekFrom,
-    path::Path,
-    pin::Pin,
-    task::{Context, Poll},
-    time::Duration,
-};
-
-use async_compression::futures::bufread::ZstdDecoder;
-use futures::{
-    io::BufReader,
-    stream::{IntoAsyncRead, MapErr},
-    AsyncRead, AsyncReadExt, AsyncSeekExt, TryStreamExt,
-};
-use pin_project_lite::pin_project;
-use thiserror::Error;
+use async_compression::tokio::bufread::ZstdDecoder;
+use futures::TryStreamExt;
+
+use indicatif::ProgressStyle;
+use log::info;
+use std::io::ErrorKind;
+use tap::Pipe;
+use tokio::io::AsyncBufReadExt;
+use tokio_util::compat::TokioAsyncReadCompatExt;
+use tokio_util::either::Either::{Left, Right};
 use url::Url;
 
-use super::https_client;
-use crate::utils::{io::ProgressBar, misc::Either};
-
-// https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md#zstandard-frames
-const ZSTD_MAGIC_HEADER: [u8; 4] = [0x28, 0xb5, 0x2f, 0xfd];
-
-#[derive(Debug, Error)]
-enum DownloadError {
-    #[error("Cannot read a file header")]
-    HeaderError,
-}
-
-pin_project! {
-    /// Holds a Reader, tracks read progress and draw a progress bar.
-    pub struct FetchProgress<R> {
-        #[pin]
-        pub inner: R,
-        pub progress_bar: ProgressBar,
-    }
-}
-
-impl<R: AsyncRead + Unpin> AsyncRead for FetchProgress<R> {
-    fn poll_read(
-        mut self: Pin<&mut Self>,
-        cx: &mut Context<'_>,
-        buf: &mut [u8],
-    ) -> Poll<std::io::Result<usize>> {
-        let r = Pin::new(&mut self.inner).poll_read(cx, buf);
-        if let Poll::Ready(Ok(size)) = r {
-            self.progress_bar.add(size as u64);
+pub struct StreamedContentReader {}
+
+impl StreamedContentReader {
+    /// This method is parsing a file-path/URL passed to it and attempts to open a stream.
+    /// Additionally, it detects whether or not the resulting stream is a zstd archive and treats
+    /// it accordingly.
+    pub async fn read(path: &str) -> anyhow::Result<Box<dyn futures::AsyncRead + Send + Unpin>> {
```

Comment:
no blocking: this should just be a function

Reply:
This has to return a `Future` one way or another, not entirely sure what this comment means. Care to elaborate?

Reply:
I think auto-merge took care of this already, but if you have any ideas for changes here - a follow-up is not going to be a problem.

Reply:
I guess you are referring to returning a `Box`d Future here.